test/core/test_task_launcher.py

```diff
@@ -88,6 +101,33 @@ def test_create_launch_expire_assignments(self):
         for assignment in launcher.assignments:
             self.assertEqual(assignment.get_status(), AssignmentState.EXPIRED)
 
+    def test_launch_assignments_with_concurrent_unit_cap(self):
+        """Initialize a launcher on a task run, then create the assignments"""
+        cap_values = [1, 2, 3, 4, 5]
+        for max_num_units in cap_values:
+            mock_data_array = self.get_mock_assignment_data_array()
+            launcher = TaskLauncher(
+                self.db,
+                self.task_run,
+                mock_data_array,
+                max_num_concurrent_units=max_num_units,
+            )
+            launcher.launched_units = LimitedDict(launcher.max_num_concurrent_units)
+            launcher.create_assignments()
+            launcher.launch_units("dummy-url:3000")
+
+            while set([u.get_status() for u in launcher.units]) != {
```

Comment:
In theory this test will loop infinitely if the system doesn't work. As such you should get a `start_time` before the loop, and have a `self.assert...` inside the loop ensuring that the current time isn't greater than some amount later than that `start_time`.